lib/puppet/util/at_fork/solaris.rb

```
@@ -0,0 +1,155 @@
+      [:ct_tmpl_activate,        Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_clear,           Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+
+      [:ct_tmpl_set_informative, Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_tmpl_set_critical,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_param,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_fatal,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+
+      [:ct_status_read,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP],
+
+      [:ct_status_get_id,        Fiddle::TYPE_INT,  Fiddle::TYPE_VOIDP],
+      [:ct_status_free,          Fiddle::TYPE_VOID, Fiddle::TYPE_VOIDP],
+    ],
+  }.each do |library, functions|
+    libhandle = Fiddle::Handle.new(library)
+
+    functions.each do |f|
+      define_method f[0], Fiddle::Function.new(libhandle[f[0].to_s], f[2..-1], f[1]).method(:call).to_proc
+    end
+  end
+
+  CTFS_PR_ROOT = File.join('', %w(system contract process))
+  CTFS_PR_TEMPLATE = File.join(CTFS_PR_ROOT, %q(template))
+  CTFS_PR_LATEST = File.join(CTFS_PR_ROOT, %q(latest))
+
+  CT_PR_PGRPONLY = 0x4
+  CT_PR_EV_HWERR = 0x20
+
+  CTD_COMMON = 0
+
+  def raise_if_error()
+    unless (e = yield) == 0
+      e = SystemCallError.new(nil, e)
+      raise e, e.message, caller
+    end
+  end
+
+  def activate_new_contract_template()
+    begin
+      tmpl = File.open(CTFS_PR_TEMPLATE, File::RDWR)
+
+      begin
+        tmpl_fd = tmpl.fileno
+
+        raise_if_error { ct_pr_tmpl_set_param(tmpl_fd, CT_PR_PGRPONLY) }
+        raise_if_error { ct_pr_tmpl_set_fatal(tmpl_fd, CT_PR_EV_HWERR) }
+        raise_if_error { ct_tmpl_set_critical(tmpl_fd, 0) }
+        raise_if_error { ct_tmpl_set_informative(tmpl_fd, CT_PR_EV_HWERR) }
+
+        raise_if_error { ct_tmpl_activate(tmpl_fd) }
+      rescue
+        tmpl.close()
+        raise
+      end
```

You could change that to just:

```ruby
ensure
  tmpl.close
end
```
So that you don't need to re-raise the exception.